Upgrade EUI to v77.2.2

1Copenut opened the pull request:
## Summary

`eui@77.1.1` ⏩ `eui@77.2.2`

Closes https://github.com/elastic/eui/issues/6724

---

## [`77.2.2`](https://github.com/elastic/eui/tree/v77.2.2)

- Updated `EuiFocusTrap` to support the `gapMode` prop configuration (now defaults to `padding`) ([#6744](https://github.com/elastic/eui/pull/6744))

**Bug fixes**

- Fixed the `scrollLock` property on `EuiFocusTrap` (and other components using `EuiFocusTrap`, such as `EuiFlyout` and `EuiModal`) to no longer block scrolling on nested portalled content, such as combobox dropdowns ([#6744](https://github.com/elastic/eui/pull/6744))

## [`77.2.1`](https://github.com/elastic/eui/tree/v77.2.1)

**Bug fixes**

- Fixed `EuiFieldNumber`'s native browser validity detection causing extra unnecessary rerenders ([#6741](https://github.com/elastic/eui/pull/6741))

## [`77.2.0`](https://github.com/elastic/eui/tree/v77.2.0)

- Updated `EuiFieldNumber` to detect native browser invalid state and show an invalid icon ([#6704](https://github.com/elastic/eui/pull/6704))
- Improved the input widths of `EuiRange` and `EuiDualRange` when `showInput={true}` to account for invalid icons ([#6704](https://github.com/elastic/eui/pull/6704))
- Improved the `isInvalid` styling of `EuiDualRange` when `showInput="inputWithPopover"` ([#6704](https://github.com/elastic/eui/pull/6704))
- Updated `EuiFormControlLayoutIcons` to render left icons in expected DOM order ([#6705](https://github.com/elastic/eui/pull/6705))
- Updated `EuiDatePickerRange`'s `isInvalid` state to match other range inputs ([#6705](https://github.com/elastic/eui/pull/6705))
- Updated `EuiSuperDatePicker`'s `isInvalid` state to match other range inputs ([#6705](https://github.com/elastic/eui/pull/6705))

**Bug fixes**

- Fixed `EuiValidatableControl` to correctly display `isInvalid` states on mount ([#6705](https://github.com/elastic/eui/pull/6705))
- Fixed an issue with `EuiSearchBar` where quoted phrases were not quoted when generating an Elasticsearch query. ([#6714](https://github.com/elastic/eui/pull/6714))

Commit:
Bumping EUI to v77.2.0.

Commit:
Adding language translations for EUI v77.2.0 bump.

Commit:
Updating snapshot tests for EUI.

Commit:
Updating unit test for new EuiDatePickerRange JSX structure.

Commit:
Revising one test from mount to shallow Enzyme method.

cee-chen commented:
@elasticmachine merge upstream

cee-chen commented:
@elasticmachine merge upstream

Commit:
Fixing CSS in query string text area.

cee-chen commented:
@elasticmachine merge upstream

cee-chen committed:
Backport EUI v77.2.1

cee-chen commented:
I can't repro the [maps FTR failure](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/124135/jobs/0187c4e2-371d-4935-9515-94fc028e96e2/artifacts/0187c4ea-4363-4284-abac-5a6c9f203afb) locally, so I'm pretty sure it's a flake 😞 

@elasticmachine merge upstream

cee-chen commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into feature/eui-v77.2.0

cee-chen commented:
One more shot for the night 🤷

cee-chen commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into feature/eui-v77.2.0

1Copenut committed:
Merge branch 'main' into feature/eui-v77.2.0

Commit:
Bumping EUI to 72.2.2 backport.

Commit:
Merging upstream main.

Commit:
package.json newline.

cee-chen committed:
Fix maps location decimal degrees showing invalid state for decimal steps

cee-chen committed:
Harden maps FTR tests

- EuiFieldNumber now has more rerenders in order to detect browser validation state, which appears to interefere with Selenium and the `setValue` API

- using `clearWithKeyboard` when clearing EuiFieldNumber fixes this problem

elasticmachine commented:
Pinging @elastic/eui-team (EUI)

elasticmachine commented:
Pinging @elastic/uptime (Team:uptime)

jughosta reviewed (approved):
Data Discovery changes LGTM 👍

nreese commented:
There are some signifiant changes to EuiFieldNumber that are not well advertised. An invalid indicator is displayed for any EuiFieldNumber that accepts a decimal and does not also set `step="any"`. For example, see the screen shot below. What are the plans to resolve these issues? Are teams expected to audit each use of EuiFieldNumber? I think we need a plan to avoid shipping 8.9 with a ton of broken EuiFieldNumber input.

For example 
<img width="200" alt="Screen Shot 2023-05-02 at 7 09 34 AM" src="https://user-images.githubusercontent.com/373691/235676019-b1816e27-51a9-4eeb-bec7-f6c4d35674e5.png">

nreese commented on x-pack/plugins/maps/public/connected_components/toolbar_overlay/set_view_control/decimal_degrees_form.tsx:
```diff
@@ -90,6 +90,7 @@ export class DecimalDegreesForm extends Component<Props, State> {
             onChange={this._onLatChange}
             isInvalid={isLatInvalid}
             data-test-subj="latitudeInput"
+            step="any" // Browsers will validate decimals as invalid otherwise
```
There are other cases where EuiFieldNumber accepts decimals and step is not set. How should these be addressed?

cee-chen replied:
Our recommendation would be for your decimal usages to set `step="any"`: https://github.com/elastic/eui/issues/5327

> There are some signifiant changes to EuiFieldNumber that are not well advertised. An invalid indicator is displayed for any EuiFieldNumber that accepts a decimal and does not also set step="any".

Just to be clear, and so we're on the same page about this behavior - `EuiFieldNumber` is not doing anything to the default **browser level** [ValidityState API](https://developer.mozilla.org/en-US/docs/Web/API/ValidityState) for decimals inputs and the `step` prop.

Browsers are the ones defaulting `step` to non-decimal 1, not EUI. You can see this [default behavior in this CodeSandbox](https://codesandbox.io/s/thirsty-forest-jqh90m?file=/index.html) when you type in `0.5` into a raw `<input type="number">` with nothing but `:invalid` styling on it. You can also see [in this example that adding `step="any"`](https://codesandbox.io/s/summer-dew-3d9inb?file=/index.html) (again, a native browser/HTML API, not EUI's) to the native input removes the `:invalid` styling.

So for all plugins in Kibana now noticing this behavior, **it was already happening** prior to this release/upgrade. EUI was already setting a [red underline](https://github.com/elastic/eui/blob/0d31625d8bc461f961293202d975e1f1e32b15a1/src/global_styling/mixins/_form.scss#L227-L229) via the `:invalid` CSS selector on all form control elements, which is visible once the blue focus underline is not present. The browser-level tooltip error message was also already occurring, users likely simply weren't hovering over it to see it (or in some browsers, e.g. Safari, tooltips are less evident than e.g., Firefox).

All this change does is add an [extra more visible icon](https://github.com/elastic/eui/issues/2017) when the native browser ValidityState flags something as `:invalid`.

So to summarize:

1. If you're seeing this issue now, you already had it before, the icon just made it more obvious
2. Devs are interacting with `EuiFieldNumber` props/attributes the exact same way you would interact with a `<input type="number">` field
3. The solution on all levels (EUI usage and native HTML usage) is for those implementing a number field to set `step="any"` on fields where you expect decimals.

There is an argument to be made here for `EuiFieldNumber` to default step to `"any"` on your behalf to make this less of a manual process for consumers. In general, EUI prefers to stick to native/browser behavior as closely as possible, but there are absolutely exceptions to be made for things like accessibility, common pain points, and developer experience. We'd be super open to hearing if you'd prefer that to be the path we take.

nreese replied:
> Just to be clear, and so we're on the same page about this behavior - EuiFieldNumber is not doing anything to the default browser level [ValidityState API](https://developer.mozilla.org/en-US/docs/Web/API/ValidityState) for decimals inputs and the step prop.

understood

> If you're seeing this issue now, you already had it before, the icon just made it more obvious

This is what concerns me. I don't think this change has been given the visibility it deserves. The icon is now visible and may confuse users as the input now displays an error icon. I just want to make sure teams understand the implications and take the time to audit their usage of EuiFieldNumber. Otherwise, there are going to be a lot of broken looking numeric inputs.

I have created https://github.com/elastic/kibana/issues/156540 to track the issue for Maps.

cee-chen replied:
Gotcha. I think there's a few options we can take here:

- Add a notice to the Kibana newsletter and EUI newsletter to try and flag this behavior more and get folks to update before 8.9 FF
- As I mentioned above, I think there's an argument to be made here for defaulting `EuiFieldNumber` to `step="any"` - this would remove any work needed on Kibana's part. Do you have any thoughts / preference on this?

cee-chen replied:
Thanks Nathan - that's a super helpful point! We'll make sure to get this change added to the Kibana newsletter for awareness.

cee-chen replied:
Just wanted to follow up on this thread - after having drilled down into the default browser behavior a bit more, we realized that the reason why this behavior wasn't obvious to users & devs beforehand is because when `step` is undefined, browsers set `:invalid` state **while typing** but then set `:valid` **after the input is blurred**. Our focus styles (blue underline) were always taking precedence over invalid styles (red underline), and then on blur, both were removed. The icon, as mentioned, is what what flags the mis-alignment in what's happening vs. what's shown to the user.

After conferring with our designers, EUI has decided to go with defaulting to `step="any"` on `EuiFieldNumber` - we consider it to be a UX improvement without significant downsides over the arguably unintuitive default browser behavior. See https://github.com/elastic/eui/pull/6760 for more details, or feel free to leave us comments there.

Thanks again for flagging this @nreese, it's seriously much appreciated.

justinkambic reviewed (approved):
Synthetics changes LGTM

TattdCodeMonkey reviewed (approved):
ent-search changes LGTM

nreese reviewed (approved):
kibana-gis changes LGTM
code review, tested in chrome

cee-chen committed:
Merge branch 'main' into feature/eui-v77.2.0

stratoula commented on src/plugins/unified_search/public/query_string_input/query_string_input.tsx:
```diff
@@ -832,8 +832,9 @@ export default class QueryStringInputUI extends PureComponent<QueryStringInputPr
               >
                 {this.forwardNewValueIfNeeded(this.getQueryString())}
               </EuiTextArea>
+              {/* EUI TODO: This will need to be fixed before the Emotion conversion */}
```
Got it! So for now we are ok, we will revisit when we convert the component to use emotion instead 👍

stratoula reviewed (approved):
Visualizations team changes LGTM!

PhilippeOberti reviewed (approved):
LGTM for the Protections Experience team!
